File: onshape_cli/api/export.py
"""Export, mass-properties, and translation helpers for Onshape."""

import asyncio
from typing import Any, Dict, List, Optional

from .client import OnshapeClient


class ExportManager:
    """Handles STL/STEP/3MF export and mass-properties queries."""

    def __init__(self, client: OnshapeClient):
        self.client = client

    async def mass_properties(
        self, document_id: str, workspace_id: str, element_id: str
    ) -> Dict[str, Any]:
        """Get mass properties (volume, mass, centroid) for a Part Studio."""
        path = (
            f"/api/v6/partstudios/d/{document_id}/w/{workspace_id}/e/{element_id}"
            "/massproperties"
        )
        return await self.client.get(path)

    async def export_stl(
        self,
        document_id: str,
        workspace_id: str,
        element_id: str,
        output_path: str,
        *,
        binary: bool = True,
        units: str = "inch",
        resolution: str = "medium",
        scale: float = 1.0,
    ) -> str:
        """Export a Part Studio to an STL file on disk (synchronous endpoint).

        Returns the output path written.
        """
        path = (
            f"/api/v6/partstudios/d/{document_id}/w/{workspace_id}/e/{element_id}/stl"
        )
        params = {
            "mode": "binary" if binary else "text",
            "units": units,
            "grouping": "true",
            "scale": scale,
            "resolution": resolution,
        }
        url = f"{self.client.base_url}{path}"
        headers = {
            "Authorization": self.client._get_auth_header(),
            "Accept": "application/vnd.onshape.v1+octet-stream",
        }
        self.client._ensure_client()
        # Onshape 307-redirects STL to a regional host (e.g. cad-usw2). httpx
        # strips the Authorization header on cross-host redirects, so follow the
        # redirect manually and re-attach auth each hop.
        resp = await self.client._client.get(
            url, params=params, headers=headers, follow_redirects=False
        )
        hops = 0
        while resp.status_code in (301, 302, 303, 307, 308) and hops < 5:
            location = resp.headers["location"]
            resp = await self.client._client.get(
                location, headers=headers, follow_redirects=False
            )
            hops += 1
        resp.raise_for_status()
        with open(output_path, "wb") as fh:
            fh.write(resp.content)
        return output_path

    async def export_translation(
        self,
        document_id: str,
        workspace_id: str,
        element_id: str,
        output_path: str,
        *,
        format_name: str = "STEP",
        poll_interval: float = 1.5,
        timeout: float = 120.0,
    ) -> str:
        """Export via the asynchronous translation API (STEP, IGES, 3MF, PARASOLID...).

        Creates a translation job, polls until DONE, then downloads the result.
        """
        # 1) Kick off the translation
        create_path = (
            f"/api/v6/partstudios/d/{document_id}/w/{workspace_id}/e/{element_id}"
            "/translations"
        )
        body = {
            "formatName": format_name,
            "storeInDocument": False,
            "flattenAssemblies": False,
        }
        job = await self.client.post(create_path, data=body)
        translation_id = job.get("id")
        if not translation_id:
            raise RuntimeError(f"Translation not started: {job}")

        # 2) Poll for completion
        elapsed = 0.0
        state = job.get("requestState", "ACTIVE")
        result: Dict[str, Any] = job
        while state == "ACTIVE" and elapsed < timeout:
            await asyncio.sleep(poll_interval)
            elapsed += poll_interval
            result = await self.client.get(f"/api/v6/translations/{translation_id}")
            state = result.get("requestState", "ACTIVE")

        if state != "DONE":
            raise RuntimeError(f"Translation failed/timeout (state={state}): {result}")

        # 3) Download the external data
        external_ids: List[str] = result.get("resultExternalDataIds", [])
        if not external_ids:
            raise RuntimeError(f"No result data: {result}")
        result_did = result.get("resultDocumentId", document_id)
        dl_path = f"/api/v6/documents/d/{result_did}/externaldata/{external_ids[0]}"
        url = f"{self.client.base_url}{dl_path}"
        headers = {"Authorization": self.client._get_auth_header()}
        self.client._ensure_client()
        resp = await self.client._client.get(url, headers=headers, follow_redirects=True)
        resp.raise_for_status()
        with open(output_path, "wb") as fh:
            fh.write(resp.content)
        return output_path

File: node/src/api/export.ts
import { writeFileSync } from "node:fs";
import { setTimeout as sleep } from "node:timers/promises";

import type { OnshapeClient } from "./client";

// Isometric camera (3x4 row-major view matrix) — a sensible default.
const ISO_VIEW = "0.707,0.707,0,0,-0.408,0.408,0.816,0,0.577,-0.577,0.577,0";

function isRecord(value: unknown): value is Record<string, any> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

/** STL/STEP/3MF export, thumbnails, shaded renders, and mass properties.
 *  Port of onshape_cli/api/export.py. */
export class ExportManager {
  constructor(private readonly client: OnshapeClient) {}

  async massProperties(documentId: string, workspaceId: string, elementId: string, configuration?: string): Promise<unknown> {
    return this.client.get(`/api/v6/partstudios/d/${documentId}/w/${workspaceId}/e/${elementId}/massproperties`, {
      configuration,
    });
  }

  async exportStl(
    documentId: string,
    workspaceId: string,
    elementId: string,
    outputPath: string,
    opts: { binary?: boolean; units?: string; resolution?: string; scale?: number; configuration?: string } = {},
  ): Promise<string> {
    const path = `/api/v6/partstudios/d/${documentId}/w/${workspaceId}/e/${elementId}/stl`;
    const { buffer } = await this.client.getBinary(
      path,
      {
        mode: opts.binary === false ? "text" : "binary",
        units: opts.units ?? "inch",
        grouping: "true",
        scale: opts.scale ?? 1.0,
        resolution: opts.resolution ?? "medium",
        configuration: opts.configuration,
      },
      "application/vnd.onshape.v1+octet-stream",
    );
    writeFileSync(outputPath, buffer);
    return outputPath;
  }

  async exportTranslation(
    documentId: string,
    workspaceId: string,
    elementId: string,
    outputPath: string,
    opts: {
      formatName?: string;
      elementKind?: string;
      configuration?: string;
      pollInterval?: number;
      timeout?: number;
    } = {},
  ): Promise<string> {
    const formatName = opts.formatName ?? "STEP";
    const elementKind = opts.elementKind ?? "partstudios";
    const pollInterval = opts.pollInterval ?? 1.5;
    const timeout = opts.timeout ?? 120.0;

    const createPath = `/api/v6/${elementKind}/d/${documentId}/w/${workspaceId}/e/${elementId}/translations`;
    const body: Record<string, unknown> = { formatName, storeInDocument: false, flattenAssemblies: false };
    if (opts.configuration) body.configuration = opts.configuration;

    const job = (await this.client.post(createPath, body)) as Record<string, any>;
    const translationId = isRecord(job) ? job.id : undefined;
    if (!translationId) throw new Error(`Translation not started: ${JSON.stringify(job)}`);

    let elapsed = 0;
    let state = (isRecord(job) && job.requestState) || "ACTIVE";
    let result: Record<string, any> = job;
    while (state === "ACTIVE" && elapsed < timeout) {
      await sleep(pollInterval * 1000);
      elapsed += pollInterval;
      result = (await this.client.get(`/api/v6/translations/${translationId}`)) as Record<string, any>;
      state = (isRecord(result) && result.requestState) || "ACTIVE";
    }
    if (state !== "DONE") throw new Error(`Translation failed/timeout (state=${state}): ${JSON.stringify(result)}`);

    const externalIds: string[] = (isRecord(result) && Array.isArray(result.resultExternalDataIds) ? result.resultExternalDataIds : []) as string[];
    if (!externalIds.length) throw new Error(`No result data: ${JSON.stringify(result)}`);
    const resultDid = (isRecord(result) && result.resultDocumentId) || documentId;
    const { buffer } = await this.client.getBinary(`/api/v6/documents/d/${resultDid}/externaldata/${externalIds[0]}`);
    writeFileSync(outputPath, buffer);
    return outputPath;
  }

  async thumbnailInfo(documentId: string, workspaceId: string, elementId: string): Promise<unknown> {
    return this.client.get(`/api/v6/thumbnails/d/${documentId}/w/${workspaceId}/e/${elementId}`);
  }

  async getThumbnail(
    documentId: string,
    workspaceId: string,
    elementId: string,
    outputPath: string,
    opts: { size?: string } = {},
  ): Promise<string> {
    const size = opts.size ?? "600x340";
    const info = (await this.thumbnailInfo(documentId, workspaceId, elementId)) as Record<string, any>;
    const sizes: Array<Record<string, any>> = isRecord(info) && Array.isArray(info.sizes) ? info.sizes : [];
    if (!sizes.length) {
      throw new Error(
        "No thumbnail available yet — Onshape renders thumbnails asynchronously; retry shortly after the element is created/edited.",
      );
    }
    const chosen = sizes.find((s) => s.size === size) ?? sizes[0];
    const href = chosen.href;
    if (!href) throw new Error(`Thumbnail entry has no href: ${JSON.stringify(chosen)}`);
    const { buffer } = await this.client.getBinaryUrl(String(href));
    writeFileSync(outputPath, buffer);
    return outputPath;
  }

  async shadedView(
    documentId: string,
    workspaceId: string,
    elementId: string,
    outputPath: string,
    opts: {
      elementKind?: string;
      width?: number;
      height?: number;
      viewMatrix?: string;
      showEdges?: boolean;
      configuration?: string;
    } = {},
  ): Promise<string> {
    const elementKind = opts.elementKind ?? "partstudios";
    const path = `/api/v6/${elementKind}/d/${documentId}/w/${workspaceId}/e/${elementId}/shadedviews`;
    const params: Record<string, string | number | boolean | undefined> = {
      viewMatrix: opts.viewMatrix ?? ISO_VIEW,
      outputWidth: opts.width ?? 600,
      outputHeight: opts.height ?? 340,
      pixelSize: 0,
    };
    if (opts.showEdges ?? true) {
      params.edges = "show";
      params.showAllParts = "true";
    }
    if (opts.configuration) params.configuration = opts.configuration;
    const resp = (await this.client.get(path, params)) as Record<string, any>;
    const images: string[] = isRecord(resp) && Array.isArray(resp.images) ? resp.images : [];
    if (!images.length) throw new Error(`No shaded image returned: ${JSON.stringify(resp)}`);
    writeFileSync(outputPath, Buffer.from(images[0], "base64"));
    return outputPath;
  }
}
